### PR TITLE
Exception Handling: Adds inner exception stacktrace in GlobalEndpointManager

### LINKED ISSUE #2590: Partial stack trace due to improperly rethrown exceptions
**Describe the bug**

The following stack trace is incomplete - we should see `ManagedIdentityCredential` here:

```
Azure.Identity.CredentialUnavailableException: ManagedIdentityCredential authentication unavailable. No Managed Identity endpoint found.
   at Microsoft.Azure.Cosmos.Routing.GlobalEndpointManager.GetAccountPropertiesHelper.GetAccountPropertiesAsync()
   at Microsoft.Azure.Cosmos.GatewayAccountReader.InitializeReaderAsync()
   at Microsoft.Azure.Cosmos.CosmosAccountServiceConfiguration.InitializeAsync()
   at Microsoft.Azure.Cosmos.DocumentClient.InitializeGatewayConfigurationReaderAsync()
   at Microsoft.Azure.Cosmos.DocumentClient.GetInitializationTaskAsync(IStoreClientFactory storeClientFactory)
   at Microsoft.Azure.Cosmos.DocumentClient.EnsureValidClientAsync(ITrace trace)
   at Microsoft.Azure.Cosmos.DocumentClient.GetCollectionCacheAsync(ITrace trace)
   at Microsoft.Azure.Cosmos.ClientContextCore.GetCachedContainerPropertiesAsync(String containerUri, ITrace trace, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.CosmosQueryClientCore.GetCachedContainerQueryPropertiesAsync(String containerLink, Nullable`1 partitionKey, ITrace trace, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.Query.Core.ExecutionContext.CosmosQueryExecutionContextFactory.TryCreateCoreContextAsync(DocumentContainer documentContainer, CosmosQueryContext cosmosQueryContext, InputParameters inputParameters, ITrace trace, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.Query.Core.AsyncLazy`1.GetValueAsync(ITrace trace, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.Query.Core.Pipeline.LazyQueryPipelineStage.MoveNextAsync(ITrace trace)
   at Microsoft.Azure.Cosmos.Query.Core.Pipeline.NameCacheStaleRetryQueryPipelineStage.MoveNextAsync(ITrace trace)
   at Microsoft.Azure.Cosmos.Query.Core.Pipeline.CatchAllQueryPipelineStage.MoveNextAsync(ITrace trace)
   at Microsoft.Azure.Cosmos.Query.QueryIterator.ReadNextAsync(ITrace trace, CancellationToken cancellationToken)
   at Microsoft.Azure.Cosmos.ClientContextCore.RunWithDiagnosticsHelperAsync[TResult](ITrace trace, Func`2 task)
   at Microsoft.Azure.Cosmos.ClientContextCore.OperationHelperWithRootTraceAsync[TResult](String operationName, RequestOptions requestOptions, Func`2 task, TraceComponent traceComponent, TraceLevel traceLevel)
```

**To Reproduce**
Run a Cosmos query using an invalid MI configuration.

**Expected behavior**
Stack trace should contain the `TokenCredential`'s stack.

**Actual behavior**
Stack trace stops at `GlobalEndpointManager.GetAccountPropertiesHelper`.

**Environment summary**
SDK Version: 3.20.0
OS Version (e.g. Windows, Linux, MacOSX): Linux

**Additional context**
I believe the issue is here (and similar lines in this file):

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/e2029f2f4854c0e4decd399c35e69ef799db9f35/Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs#L164

This should use `ExceptionDispatchInfo.Capture(this.NonRetriableException).Throw()` (or capture using a EDI to begin with).

I did a quick code search and this line might also suffer from the same issue:

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/e2029f2f4854c0e4decd399c35e69ef799db9f35/Microsoft.Azure.Cosmos.Encryption/src/Mirrored/IRetryPolicy.cs#L53

### PR DESCRIPTION
**Pull request by sourabh1007**
## Description

Right now if any exceptions occurs while getting account details in globalEndpointManager, It gets rethrown without inner exception stacktrace.
So, Adding inner exception stack trace while rethrowing exception in GlobalEndpointManager. 

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)

## Closing issues

Closes #2590

### CODE REVIEW

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientTests.cs [hunk 1]
```diff
@@ -101,6 +101,31 @@ public void Builder_InvalidEndpointAndKey(string endpoint, string key)
             new CosmosClientBuilder(endpoint, key);
         }
 
+        [DataTestMethod]
+        [DataRow(AccountEndpoint, "425Mcv8CXQqzRNCgFNjIhT424GK88ckJvASowTnq15Vt8LeahXTcN5wt3342vQ==")]
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientTests.cs:105]: Remove the key. You can not store keys in plain text. Use the method all the other tests are doing by using the TestCommon method to create the client.

File: Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs [hunk 2]
```diff
@@ -183,7 +184,7 @@ public async Task<AccountProperties> GetAccountPropertiesAsync()
 
                     if (this.NonRetriableException != null)
                     {
-                        throw this.NonRetriableException;
+                        ExceptionDispatchInfo.Capture(this.NonRetriableException).Throw();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs:187]: What about line 200 that throws an exception?

File: Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs [hunk 3]
```diff
@@ -161,7 +162,7 @@ public async Task<AccountProperties> GetAccountPropertiesAsync()
 
                 if (this.NonRetriableException != null)
                 {
-                    throw this.NonRetriableException;
+                    ExceptionDispatchInfo.Capture(this.NonRetriableException).Throw();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs:165]: Would it be better to add this to the GetAndUpdateAccountPropertiesAsync method instead?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientTests.cs (hunk 1 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientTests.cs:105]: In this test class, Keys are already hardcoded (refer. other test cases)..and anyway it is invalid key.

File: Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs (hunk 3 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Routing/GlobalEndpointManager.cs:165]: Why do you want to do it in GetAndUpdateAccountPropertiesAsync(). According to me this problem occurs only when we have Tasks involved, that is where we are loosing exception thrown in the task. In this function we are just collecting exceptions. I don't want to change this implementation by throwing exception there.

**Review by j82w — APPROVED** [review]

**Pull request merged**